## Загрузка и подготовка данных

In [90]:
import pandas as pd
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error
import matplotlib.pyplot as plt
from numpy.random import RandomState
from scipy import stats as st
import numpy as np

In [91]:
data1 = pd.read_csv('/datasets/geo_data_0.csv')
data2 = pd.read_csv('/datasets/geo_data_1.csv')
data3 = pd.read_csv('/datasets/geo_data_2.csv')
display(data1)
data1.info()

id        f0        f1        f2     product
0      txEyH  0.705745 -0.497823  1.221170  105.280062
1      2acmU  1.334711 -0.340164  4.365080   73.037750
2      409Wp  1.022732  0.151990  1.419926   85.265647
3      iJLyR -0.032172  0.139033  2.978566  168.620776
4      Xdl7t  1.988431  0.155413  4.751769  154.036647
...      ...       ...       ...       ...         ...
99995  DLsed  0.971957  0.370953  6.075346  110.744026
99996  QKivN  1.392429 -0.382606  1.273912  122.346843
99997  3rnvd  1.029585  0.018787 -1.348308   64.375443
99998  7kl59  0.998163 -0.528582  1.583869   74.040764
99999  1CWhH  1.764754 -0.266417  5.722849  149.633246

[100000 rows x 5 columns]

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   id       100000 non-null  object 
 1   f0       100000 non-null  float64
 2   f1       100000 non-null  float64
 3   f2       100000 non-null  float64
 4   product  100000 non-null  float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB


In [92]:
display(data2)
data2.info()

id         f0         f1        f2     product
0      kBEdx -15.001348  -8.276000 -0.005876    3.179103
1      62mP7  14.272088  -3.475083  0.999183   26.953261
2      vyE1P   6.263187  -5.948386  5.001160  134.766305
3      KcrkZ -13.081196 -11.506057  4.999415  137.945408
4      AHL4O  12.702195  -8.147433  5.004363  134.766305
...      ...        ...        ...       ...         ...
99995  QywKC   9.535637  -6.878139  1.998296   53.906522
99996  ptvty -10.160631 -12.558096  5.005581  137.945408
99997  09gWa  -7.378891  -3.084104  4.998651  137.945408
99998  rqwUm   0.665714  -6.152593  1.000146   30.132364
99999  relB0  -3.426139  -7.794274 -0.003299    3.179103

[100000 rows x 5 columns]

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   id       100000 non-null  object 
 1   f0       100000 non-null  float64
 2   f1       100000 non-null  float64
 3   f2       100000 non-null  float64
 4   product  100000 non-null  float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB


In [93]:
display(data3)
data3.info()

id        f0        f1        f2     product
0      fwXo0 -1.146987  0.963328 -0.828965   27.758673
1      WJtFt  0.262778  0.269839 -2.530187   56.069697
2      ovLUW  0.194587  0.289035 -5.586433   62.871910
3      q6cA6  2.236060 -0.553760  0.930038  114.572842
4      WPMUX -0.515993  1.716266  5.899011  149.600746
...      ...       ...       ...       ...         ...
99995  4GxBu -1.777037  1.125220  6.263374  172.327046
99996  YKFjq -1.261523 -0.894828  2.524545  138.748846
99997  tKPY3 -1.199934 -2.957637  5.219411  157.080080
99998  nmxp2 -2.419896  2.417221 -5.548444   51.795253
99999  V9kWn -2.551421 -2.025625  6.090891  102.775767

[100000 rows x 5 columns]

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   id       100000 non-null  object 
 1   f0       100000 non-null  float64
 2   f1       100000 non-null  float64
 3   f2       100000 non-null  float64
 4   product  100000 non-null  float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB


В данных нет пропусков. Не понятно, что означают метрики. Но, по условию задачи, это не важно.

In [94]:
print(data1.duplicated().sum())
print(data2.duplicated().sum())
print(data3.duplicated().sum())

0
0
0


Дубликатов не обнаружено. Удалим стобцы с id. Для обучнения модели он не нужен

In [95]:
data11 = data1.drop(['id'], axis=1)
data22 = data2.drop(['id'], axis=1)
data33 = data3.drop(['id'], axis=1)

In [96]:
data33

f0        f1        f2     product
0     -1.146987  0.963328 -0.828965   27.758673
1      0.262778  0.269839 -2.530187   56.069697
2      0.194587  0.289035 -5.586433   62.871910
3      2.236060 -0.553760  0.930038  114.572842
4     -0.515993  1.716266  5.899011  149.600746
...         ...       ...       ...         ...
99995 -1.777037  1.125220  6.263374  172.327046
99996 -1.261523 -0.894828  2.524545  138.748846
99997 -1.199934 -2.957637  5.219411  157.080080
99998 -2.419896  2.417221 -5.548444   51.795253
99999 -2.551421 -2.025625  6.090891  102.775767

[100000 rows x 4 columns]

In [97]:
data11 = pd.get_dummies(data11)

In [98]:
data11

f0        f1        f2     product
0      0.705745 -0.497823  1.221170  105.280062
1      1.334711 -0.340164  4.365080   73.037750
2      1.022732  0.151990  1.419926   85.265647
3     -0.032172  0.139033  2.978566  168.620776
4      1.988431  0.155413  4.751769  154.036647
...         ...       ...       ...         ...
99995  0.971957  0.370953  6.075346  110.744026
99996  1.392429 -0.382606  1.273912  122.346843
99997  1.029585  0.018787 -1.348308   64.375443
99998  0.998163 -0.528582  1.583869   74.040764
99999  1.764754 -0.266417  5.722849  149.633246

[100000 rows x 4 columns]

Предоставленные данные уже закодированы.

## Обучение и проверка модели

Согласно условиям задачи, для обучения модели подходит только линейная регрессия (остальные — недостаточно предсказуемые).

### Для 1 региона

In [99]:
features_1 = data11.drop(['product'], axis=1)
target_1 = data11['product']

In [100]:
features_train_1, features_valid_1, target_train_1, target_valid_1 = train_test_split(
    features_1, target_1, test_size=0.25, random_state=14)

In [101]:
target_valid_1 = target_valid_1.reset_index(drop=True)
model_1 = LinearRegression()
model_1.fit(features_train_1, target_train_1)
prediction_valid_1 = model_1.predict(features_valid_1)
result_1 = mean_squared_error(target_valid_1, prediction_valid_1)**0.5
print('RMSE модели 1:', result_1)
print('Cредний запас предсказанного сырья', prediction_valid_1.mean())

RMSE модели 1: 37.64172219212394
Cредний запас предсказанного сырья 92.56331572920227


### Для 2 региона

In [102]:
features_2 = data22.drop(['product'], axis=1)
target_2 = data22['product']
features_train_2, features_valid_2, target_train_2, target_valid_2 = train_test_split(
    features_2, target_2, test_size=0.25, random_state=14)
target_valid_2 = target_valid_2.reset_index(drop=True)
model_2 = LinearRegression()
model_2.fit(features_train_2, target_train_2)
prediction_valid_2 = model_2.predict(features_valid_2)
result_2 = mean_squared_error(target_valid_2, prediction_valid_2)**0.5
print('RMSE модели 2:', result_2)
print('Cредний запас предсказанного сырья', prediction_valid_2.mean())

RMSE модели 2: 0.8925967602874482
Cредний запас предсказанного сырья 68.89452655451875


### Для 3 региона


In [103]:
features_3 = data33.drop(['product'], axis=1)
target_3 = data33['product']
features_train_3, features_valid_3, target_train_3, target_valid_3 = train_test_split(
    features_3, target_3, test_size=0.25, random_state=14)
target_valid_3 = target_valid_3.reset_index(drop=True)
model_3 = LinearRegression()
model_3.fit(features_train_3, target_train_3)
prediction_valid_3 = model_3.predict(features_valid_3)
result_3 = mean_squared_error(target_valid_3, prediction_valid_3)**0.5
print('RMSE модели 3:', result_3)
print('Cредний запас предсказанного сырья', prediction_valid_3.mean())


RMSE модели 3: 39.824525086273866
Cредний запас предсказанного сырья 95.04162515658933


In [104]:
rmse_mean = pd.DataFrame(data=[['1', result_1, prediction_valid_1.mean()], 
                               ['2', result_2, prediction_valid_2.mean()], 
                               ['3', result_3, prediction_valid_3.mean()]], 
                         columns=[['Регион', 'RMSE модели', 'Cредний запас предсказанного сырья']])
print('Получили следующие результаты после обучения моделей:')
rmse_mean

Получили следующие результаты после обучения моделей:


Регион RMSE модели Cредний запас предсказанного сырья
0      1   37.641722                          92.563316
1      2    0.892597                          68.894527
2      3   39.824525                          95.041625

Получается, что лучшее предсказание у нас получилось по второму региону, RMSE которого равна 0.892597. Но во втором регионе и самый минимальный средний запас сырья. Однако, хоть у других регионов  средний запас больше, но и RMSE больше. Значит, например, для 1 региона, мы можем ошибаться на +-37 единиц. Значит, что запас сырья может быть как 92, так и 92-37=55. 

## Подготовка к расчёту прибыли

Ключевые значения для расчётов:

In [120]:
BOREHOLES= 500 # исследуемое количество скважин при разведке 
BEST_BOREHOLES= 200 # количество лучших скважин
BUDGET = 10000000000 # бюджет на разработку скважин в регионе 
PROFIT = 450000 # доход с каждой единицы продукта
MIN_PROFIT_FOR_REGION = BUDGET/PROFIT # минимальный объем добычи на регион для 200 скважин
MIN_PROFIT_FOR_BOREHOLE = MIN_PROFIT_FOR_REGION/BEST_BOREHOLES # минимальный объем добычи на регион для 200 скважин
MIN_PROFIT_FOR_BOREHOLE

111.11111111111111

Чтобы окупить бюджет на расходы (10 млрд.), каждая из 200 скважин должна вырабатывать минимум 111,112 тысяч баррелей. Среднее для каждого из регионов меньше, чем минимальный объем добычи. Но мы сравниваем со средним. А мы будем выбирать 200 лучших сважин, объем которых будет выше, чем средний объем по региону.

## Расчёт прибыли и рисков 

In [121]:
def money(target, prediction):
    prediction_valid = prediction.sort_values(ascending=False)
    selected = target[prediction_valid.index][:200]
    global PROFIT
    global BUDGET
    return (selected.sum()*profit-budget)/1000000000

In [122]:
state = RandomState(14)

**Регион 1**

In [123]:
prediction_valid_1 = pd.Series(prediction_valid_1)
values_1 = []
for i in range(1000):
    target = target_valid_1.sample(500, replace=True, random_state=state)   
    probs = prediction_valid_1[target.index]
    values_1.append(money(target, probs))
values_1 = pd.Series(values_1)
values_1.mean()

0.46710561923987265

95%-й Доверительный интервал для региона 1:

In [124]:
lower_1 = values_1.quantile(0.025)
upper_1 = values_1.quantile(0.975)
print(lower_1)
print(upper_1)

-0.05790326671323156
1.017560704948621


**Регион 2**

In [125]:
prediction_valid_2 = pd.Series(prediction_valid_2)
values_2 = []
for i in range(1000):
    target = target_valid_2.sample(500, replace=True, random_state=state)
    probs = prediction_valid_2[target.index]
    values_2.append(money(target, probs))
values_2 = pd.Series(values_2)
values_2.mean()

0.48448940048518757

95%-й Доверительный интервал для региона 2:

In [126]:
lower_2 = values_2.quantile(0.025)
upper_2 = values_2.quantile(0.975)
print(lower_2)
print(upper_2)

0.07240151090591998
0.8929323798418827


**Регион 3**

In [127]:
values_3 = []
prediction_valid_3 = pd.Series(prediction_valid_3)
for i in range(1000):
    target = target_valid_3.sample(500, replace=True, random_state=state)
    probs = prediction_valid_3[target.index]
    values_3.append(money(target, probs))
values_3 = pd.Series(values_3)
values_3.mean()

0.4213813881289648

95%-й Доверительный интервал для региона 3:

In [128]:
lower_3 = values_3.quantile(0.025)
upper_3 = values_3.quantile(0.975)
print(lower_3)
print(upper_3)

-0.1326805297204353
0.9608777101621837


In [129]:
region_money = pd.DataFrame(data=[['1', values_1.mean(), lower_1, upper_1], 
                               ['2', values_2.mean(), lower_2, upper_2], 
                               ['3', values_3.mean(), lower_3, upper_3]], 
                         columns=[['Регион', 'Средняя прибыль с одной скважины, млрд. руб.', 
                                   'Доверительный интервал начало', 'Доверительный интервал конец']])
region_money

Регион Средняя прибыль с одной скважины, млрд. руб.  \
0      1                                     0.467106   
1      2                                     0.484489   
2      3                                     0.421381   

  Доверительный интервал начало Доверительный интервал конец  
0                     -0.057903                     1.017561  
1                      0.072402                     0.892932  
2                     -0.132681                     0.960878

Тут в глаза бросается то, что у второго региона интервал начинается с положительной выручки, а у других регионов нет Значит, что в минимум в 95% случаев при работе во втором регионе не будет убытка.


**Рассчет рисков**

In [130]:
print('Для региона 1 риск убытков равен', (values_1 < 0).mean()*100, '%')
print('Для региона 2 риск убытков равен', (values_2 < 0).mean()*100, '%')
print('Для региона 3 риск убытков равен', (values_3 < 0).mean()*100, '%')

Для региона 1 риск убытков равен 3.5000000000000004 %
Для региона 2 риск убытков равен 0.8 %
Для региона 3 риск убытков равен 6.3 %


Для разработки скважин оптималный выбор это Регион 2.

RMSE модели, обученной по данным второго региона минимальна : 0.892597.

Несмотря на то, у этого региона самый минимальный средний предсказанный объем на скважину, после того, как мы 1000 раз выбрали 500 случайных скважин, 200 лучших скважин из этих 500 показали среднюю прибыль около 480 млн. руб. 

Данные для этого региона лежат в 95% доверительном интервале. Риск убытков равен 0,8% - минимальный среди трех регионов.

Если в этом регионе выберут 500 случайных скважин, а затем 200 лучших из этих 500, то с вероятность 99,2% это будет успешный выбор, который принесет прибыль.


